ssl/quic/quic_srtm.c

```diff
@@ -0,0 +1,576 @@
+ * If head is non-NULL, writes the head of the relevant opaque list to *head if
+ * there is one.
+ * If prev is non-NULL, writes the previous node to *prev or NULL if it is
+ * the first item.
+ */
+static SRTM_ITEM *srtm_find(QUIC_SRTM *srtm, void *opaque, uint64_t seq_num,
+                            SRTM_ITEM **head_p, SRTM_ITEM **prev_p)
+{
+    SRTM_ITEM key, *item = NULL, *prev = NULL;
+
+    key.opaque  = opaque;
+
+    item = lh_SRTM_ITEM_retrieve(srtm->items_fwd, &key);
+    if (head_p != NULL)
+        *head_p = item;
+
+    for (; item != NULL; prev = item, item = item->next_by_seq_num)
+        if (item->seq_num == seq_num) {
+            break;
+        } else if (item->seq_num < seq_num) {
+            /*
+             * List is sorted in descending order so there can't be any match
+             * after this.
+             */
+            item = NULL;
+            break;
+        }
+
+    if (prev_p != NULL)
+        *prev_p = prev;
+
+    return item;
+}
+
+/*
+ * Inserts a SRTM_ITEM into the singly-linked by-sequence-number linked list.
+ * The new head pointer is written to *new_head (which may or may not be
+ * unchanged).
+ */
+static void sorted_insert_seq_num(SRTM_ITEM *head, SRTM_ITEM *item, SRTM_ITEM **new_head)
+{
+    uint64_t seq_num = item->seq_num;
+    SRTM_ITEM *cur = head, **fixup = new_head;
+
+    *new_head = head;
+
+    while (cur != NULL && cur->seq_num > seq_num) {
+        fixup = &cur->next_by_seq_num;
+        cur = cur->next_by_seq_num;
+    }
+
+    item->next_by_seq_num = *fixup;
+    *fixup = item;
+}
+
+/*
+ * Inserts a SRTM_ITEM into the singly-linked by-SRT list.
+ * The new head pointer is written to *new_head (which may or may not be
+ * unchanged).
+ */
+static void sorted_insert_srt(SRTM_ITEM *head, SRTM_ITEM *item, SRTM_ITEM **new_head)
+{
+    uintptr_t opaque = (uintptr_t)item->opaque;
+    SRTM_ITEM *cur = head, **fixup = new_head;
+
+    *new_head = head;
+
+    while (cur != NULL && (uintptr_t)cur->opaque > opaque) {
+        fixup = &cur->next_by_srt_blinded;
+        cur = cur->next_by_srt_blinded;
+    }
+
+    item->next_by_srt_blinded = *fixup;
+    *fixup = item;
+}
+
+/*
+ * Computes the blinded SRT value used for internal lookup for side channel
+ * mitigation purposes. We compute this once as a cached value when an SRTM_ITEM
+ * is formed.
+ */
+static int srtm_compute_blinded(QUIC_SRTM *srtm, SRTM_ITEM *item,
+                                const QUIC_STATELESS_RESET_TOKEN *token)
+{
+    size_t outl = 0;
+
+    if (!EVP_MAC_init(srtm->blind_ctx, NULL, 0, NULL))
+        return 0;
+
+    if (!EVP_MAC_update(srtm->blind_ctx, (const unsigned char *)token,
+                        sizeof(*token)))
+        return 0;
+
+    if (!EVP_MAC_final(srtm->blind_ctx, item->srt_blinded,
+                       &outl, sizeof(item->srt_blinded))
+        || outl != sizeof(item->srt_blinded))
+        return 0;
+
+    return 1;
+}
```

Comment:
We should really use AES-128-ECB here with a random key generated on SRTM init as the block size is the same as the SRT length. It is faster and guaranteed to be a permutation.

Reply:
Makes sense. OK.